File: 20250321/leetcode2680. 最大或值/main.cpp
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  long long maximumOr(vector<int> &nums, int k) {
    int n = nums.size();
    long long mutil = ((long long)1) << k;
    vector<long long> suffix(n + 1);
    suffix[n] = 0;
    for (int i = n - 1; i >= 0; i--) {
      suffix[i] = suffix[i + 1] | (long long)nums[i];
    }
    long long pre = 0;
    long long res = 0;
    for (int i = 0; i < n; i++) {
      long long cur = nums[i] * mutil | pre | suffix[i + 1];
      res = max(res, cur);
      pre = pre | nums[i];
    }
    return res;
  }
  long long maximumOr2(vector<int> &nums, int k) {
    int n = nums.size();
    // 计算每个数字的或值和1数量大于1的并值
    long long multiBits = 0;
    long long orSum = 0;
    for (int x : nums) {
      long long lx = x;
      multiBits = multiBits | (lx & orSum);
      orSum = orSum | lx;
    }
    long long res = 0;
    long long lk = k;
    for (int x : nums) {
      long long lx = x;
      // orSum与lx的异或结果：
      // lx中为0的，orSum中为1的 => 1，少了lx，这个位也是1，正确
      // lx中为0的，orSum中为0的 => 0，少了lx，这个位也是0，正确，
      // lx中为1的，orSum中为1的 => 0,
      // 少了lx，这个位由multiBits决定，数量大于1则为1，只有一个则为0
      // lx中为1的，orSum中为0的 => 不存在这种情况
      // 所以(orSum ^ lx) | multiBits 就是除了x剩下的
      res = max(res, (orSum ^ lx) | multiBits | (lx << lk));
    }
    return res;
  }
};

int main(void) {
  Solution solution;
  vector<int> nums = {98545425, 54785416, 85745865, 98588888,  65475895,
                      25895555, 79899999, 65847589, 100000000, 25684759};
  cout << solution.maximumOr(nums, 15) << endl;
  cout << solution.maximumOr2(nums, 15) << endl;
  return 0;
}

File: 20250321/leetcode2680. 最大或值/main.js
// 给你一个下标从 0 开始长度为 n 的整数数组 nums 和一个整数 k 。每一次操作中，你可以选择一个数并将它乘 2 。

// 你最多可以进行 k 次操作，请你返回 nums[0] | nums[1] | ... | nums[n - 1] 的最大值。

// a | b 表示两个整数 a 和 b 的 按位或 运算。

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maximumOr = function (nums, k) {
  const n = nums.length;
  // 必定只有一个数会被连续左移k位
  // 遍历nums，将其中一个数左移k位，前缀或可以在遍历过程中得知，后缀或使用后缀和的方式存储在数组中
  const suffix = new Array(n + 1).fill(0n);
  for (let i = n - 1; i >= 0; i--) {
    suffix[i] = suffix[i + 1] | BigInt(nums[i]);
  }
  let pre = 0n;
  let res = 0;
  const mutil = BigInt(2 ** k);
  for (let i = 0; i < n; i++) {
    let cur = (mutil * BigInt(nums[i])) | suffix[i + 1] | pre;
    res = Math.max(res, Number(cur));
    pre = pre | BigInt(nums[i]);
  }
  return res;
};

// 12 => 1100   => 11000
// 9  => 1001   => 10010
//
